### PR TITLE
re-enable routing for egress rules

### PR DESCRIPTION
**Pull request by rshriram**
**What this PR does / why we need it**:
This PR enables routing support for Egress rules that were introduced earlier. In addition, it fixes two more bugs: allowing wildcard hostnames in ingress, and processing egress rules when HTTP_PROXY is set (this was missing as well).
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
#1044 
**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

cc @vadimeisenbergibm 

In future (during 0.3, we can revisit cleaning up this entire piece - where we can source egress rules as a registryadapter). Right now, the focus is on fixing the issue.

### CODE REVIEW

File: proxy/envoy/config.go [hunk 1]
```diff
@@ -802,3 +736,66 @@ func buildMgmtPortListeners(mesh *proxyconfig.MeshConfig, managementPorts model.
...
+	}
+
+	for _, rule := range egressRules {
+		for _, port := range rule.Ports {
+			protocol := model.Protocol(strings.ToUpper(port.Protocol))
+			intPort := int(port.Port)
+			servicePort := &model.Port{Name: "external-traffic-port", Port: intPort, Protocol: protocol}
+			destination := rule.Destination.Service
+			svc := &model.Service{Hostname: destination}
+			routes := buildDestinationHTTPRoutes(svc, servicePort, instances, config)
```
Comment (kyessenov) [inline proxy/envoy/config.go]: How does this work with labelled endpoint routing? I assume the rule would have externhost in the destination service field. Is the weighted LB not supported?

File: proxy/envoy/config.go [hunk 2]
```diff
@@ -802,3 +736,66 @@ func buildMgmtPortListeners(mesh *proxyconfig.MeshConfig, managementPorts model.
...
+	config model.IstioConfigStore, httpConfigs HTTPRouteConfigs) HTTPRouteConfigs {
+
+	// Convert all egress rules into services and add route rules if any
+	egressRules, errs := model.RejectConflictingEgressRules(config.EgressRules())
+
+	if errs != nil {
+		glog.Warningf("Rejected rules: %v", errs)
+	}
+
+	for _, rule := range egressRules {
```
Comment (kyessenov) [inline proxy/envoy/config.go]: Are these rules sorted?

File: proxy/envoy/config.go (hunk 1 above)
Comment (rshriram) [reply proxy/envoy/config.go]: Sorry I missed this. No label based routing/weighted LB. What I wanted to support was DP, retries, timeouts and fault injection.

**Comment by rshriram** [comment]
@kyessenov ptal

File: proxy/envoy/testdata/rds-egress-rule-timeout-ssl.json.golden [hunk 3]
```diff
@@ -0,0 +1,26 @@
+{
+  "virtual_hosts": [
```
Comment (rshriram) [inline proxy/envoy/testdata/rds-egress-rule-timeout-ssl.json.golden]: Not good. 443 missing.

File: proxy/envoy/testdata/cds-circuit-breaker.json.golden [hunk 4]
```diff
@@ -103,7 +103,7 @@
     "service_name": "world.default.svc.cluster.local|http-status|version=v0",
     "connect_timeout_ms": 1000,
     "type": "sds",
-    "lb_type": "random",
+    "lb_type": "round_robin",
```
Comment (kyessenov) [inline proxy/envoy/testdata/cds-circuit-breaker.json.golden]: Why did this change? Not seeing the cause for this...

**Comment by rshriram** [comment]
Don't merge yet.

File: proxy/envoy/testdata/lds-v0-egressrule-auth.json.golden [hunk 5]
```diff
@@ -0,0 +1,610 @@
+{
```
Comment (kyessenov) [inline proxy/envoy/testdata/lds-v0-egressrule-auth.json.golden]: This seems to be identical to lds-v0-auth.json.golden. Might be worth removing to make it clear.

File: proxy/envoy/testdata/lds-v0-egressrule-nomixer.json.golden [hunk 6]
```diff
@@ -0,0 +1,462 @@
+{
```
Comment (kyessenov) [inline proxy/envoy/testdata/lds-v0-egressrule-nomixer.json.golden]: Same. Please compare to non egressrule to make sure egress rules do not affect LDS.

File: proxy/envoy/testdata/egress-cb-policy.yaml.golden [hunk 7]
```diff
@@ -0,0 +1,22 @@
+type: destination-policy
```
Comment (kyessenov) [inline proxy/envoy/testdata/egress-cb-policy.yaml.golden]: We need CDS output corresponding to clusters in RDS (especially for auth and mixer).

File: proxy/envoy/testdata/rds-egress-rule-timeout-http.json.golden [hunk 8]
```diff
@@ -0,0 +1,91 @@
+{
+  "virtual_hosts": [
+   {
+    "name": "*.google.com:80",
+    "domains": [
+     "*.google.com",
+     "*.google.com:80"
+    ],
+    "routes": [
```
Comment (kyessenov) [inline proxy/envoy/testdata/rds-egress-rule-timeout-http.json.golden]: Need to set OpaqueConfig for mixer. By defaults, it forwards attributes from source, which should not happen for out-of-cluster destinations.

File: proxy/envoy/testdata/egress-cb-policy.yaml.golden (hunk 7 above)
Comment (rshriram) [reply proxy/envoy/testdata/egress-cb-policy.yaml.golden]: Yes I know. Don't know why it's missing. I think I didn't add those test cases

**Comment by kyessenov** [comment]
This looks to be pretty invasive. Should we park this for later?

**Comment by rshriram** [comment]
Its all golden file updates. 42 goldens!! I can confine the changes in discovery_test.go if need be.

**Comment by rshriram** [comment]
Pr has been broken into smaller ones. Will do another PR to enable routing

**Pull request closed without merging**